Define constants for semantic convention attribute names

kbrockhoff opened the pull request:
Defines String constants for all span attribute names listed in the semantic convention specifications.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/SemanticConventions.java:
```diff
@@ -0,0 +1,91 @@
+ *     href="https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md">Semantic
+ *     Conventions</a>
+ */
+public final class SemanticConventions {
```
Do we commit on these values to be 18-24 months backwards compatible? Maybe we should have a contrib package for this. 

@tedsuo @yurishkuro what is your experience with this?

bogdandrutu replied:
Ping @tedsuo and @yurishkuro

Oberon00 replied:
I think we should start treating semantic conventions more like an API definition (a contract between instrumentation writers and the back-end), so I'd love if we could guarantee at least 18-24 months compatibility.
Still, like typed spans (#778), a contrib package seems fine (I'd rather have typed spans in the API than the string constants).

tylerbenson commented on api/src/main/java/io/opentelemetry/trace/SemanticConventions.java:
```diff
@@ -0,0 +1,91 @@
+ *     href="https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md">Semantic
+ *     Conventions</a>
+ */
+public final class SemanticConventions {
```
Since these all apply to attributes, could we rename the class to `SemanticAttributes`?

bogdandrutu replied:
I think it is a good suggestion.

carlosalberto replied:
I like `SemanticAttributes` more as well.

kbrockhoff replied:
Changed

Oberon00 commented:
Could we do something like OpenTelemetry's "Tag setters" instead? https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/tag/Tags.java

With that, you'd write something like `Tags.HTTP_STATUS.set(span, 200)` and the set method would only accept integers at compile time.

yurishkuro commented on api/src/main/java/io/opentelemetry/trace/SemanticAttributes.java:
```diff
@@ -0,0 +1,95 @@
+  /** Actual/assumed role the client is making the request under. */
+  public static final String ENDUSER_ROLE = "enduser.role";
+  /** Scopes or granted authorities the client currently possesses. */
+  public static final String ENDUSER_SCOPE = "enduser.scope";
```
Please use full descriptions from the spec.

kbrockhoff replied:
Fixed.

kbrockhoff commented:
> Could we do something like OpenTelemetry's "Tag setters" instead? https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/tag/Tags.java
> 
> With that, you'd write something like `Tags.HTTP_STATUS.set(span, 200)` and the set method would only accept integers at compile time.

This is a good idea. I have implemented and moved into separate package.

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/AbstractAttribute.java:
```diff
@@ -0,0 +1,45 @@
+ * @param <T> the attribute value type
+ */
+@Immutable
+public abstract class AbstractAttribute<T> implements Attribute<T> {
```
I suggest naming these `*AttributeSetter` instead of just `*Attribute` to avoid confusion with already existing classes like `io.opentelemetry.trace.AttributeValue`.

kbrockhoff replied:
Changed to AttributeSetter

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttribute.java:
```diff
@@ -0,0 +1,44 @@
+    if (value != null) {
+      span.setAttribute(key(), value);
+    } else {
+      span.setAttribute(key(), (String) null);
```
This case looks really awkward. Why would we convert a boolean to a string? Isn't setting a null string equivalent to not calling `setAttribute` as opposed to clearing the attribute anyway?

kbrockhoff replied:
Changed

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/IntOrStringAttribute.java:
```diff
@@ -0,0 +1,61 @@
+ * or strings in most applications.
+ */
+@Immutable
+public class IntOrStringAttribute extends AbstractAttribute<Integer> {
```
I don't like this class. Also, from the name I would have expected it not to parse the string but leave it as-is.

kbrockhoff replied:
This was copied from OpenTracing based on the assumption that it was added there because of need

Oberon00 replied:
I challenge that assumption ;) Or rather, I have an alternative idea, see https://github.com/open-telemetry/opentelemetry-java/pull/758#discussion_r389513439

kbrockhoff replied:
Changed

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/IntOrStringAttribute.java:
```diff
@@ -0,0 +1,61 @@
+   */
+  public void set(Span span, @Nullable String value) {
+    if (value != null) {
+      span.setAttribute(key(), Long.valueOf(value));
```
This may throw an exception for invalid input which violates the specification's error handling guidelines. Also, since this input will most likely come directly from the monitored code, it is even more dangerous.

kbrockhoff replied:
Good catch. I have added try catch block

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/IntOrStringAttributeSetter.java:
```diff
@@ -0,0 +1,65 @@
+ * or strings in most applications.
+ */
+@Immutable
+public class IntOrStringAttributeSetter extends AbstractAttributeSetter<Integer> {
```
What about this: We remove this class but instead give LongAttributeSetter two methods:

* trySetParsed(String) which tries to parse the argument and does nothing on failure.
* setParsedOrRaw(String) which tries to parse the argument and sets the string as-is on failure
* remove set(String) from to avoid *accidentally* passing a string

I would think that this is much cleaner than having this IntOrStringAttributeSetter  where it is not clear what set does on failure.

Oberon00 commented on api/src/main/java/io/opentelemetry/trace/attributes/AttributeSetter.java:
```diff
@@ -0,0 +1,39 @@
+   * @param span the span to add the attribute to
+   * @param value the value for this attribute or <code>null</code> to remove attribute from span
+   */
+  void set(Span span, @Nullable T value);
```
I think there is no benefit in having this in the interface. In fact there is a drawback, because it forces as to use (nullable) Long instead of just primitive `long`. In fact, I'd vote for removing this interface entirely (actually, we could even remove the abstract base class since the achieved code sharing is extremely minimal: We'd write `private final String key` instead of `extends AbstractAttributeSetter<Foo>` and `this.key = key` instead of `super(key)` and that's about it.)

Oberon00 commented:
@kbrockhoff I would really love to have this feature merged. What are your plans for this PR?

kbrockhoff commented:
> @kbrockhoff I would really love to have this feature merged. What are your plans for this PR?

I will get recommended changes done within the next couple of days.

Oberon00 commented on api/src/test/java/io/opentelemetry/trace/attributes/SemanticAttributesTest.java:
```diff
@@ -0,0 +1,85 @@
+        ((LongAttributeSetter) attribute).trySetParsed(span, null);
+        ((LongAttributeSetter) attribute).setParsedOrRaw(span, "42");
+        ((LongAttributeSetter) attribute).setParsedOrRaw(span, "BAD");
+        ((LongAttributeSetter) attribute).setParsedOrRaw(span, null);
```
```suggestion
        final LongAttributeSetter longAttribute = (LongAttributeSetter) attribute:
        keys.add(longAttribute .key());
        longAttribute.set(span, 42L);
        longAttribute.trySetParsed(span, "42");
        longAttribute.trySetParsed(span, "BAD");
        longAttribute.trySetParsed(span, null);
        longAttribute.setParsedOrRaw(span, "42");
        longAttribute.setParsedOrRaw(span, "BAD");
        longAttribute.setParsedOrRaw(span, null);
```

and so on

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,83 @@
+    } else {
+      span.setAttribute(key(), value);
+    }
+  }
```
Do we really need these methods now? I would suggest to stay only with `set` for the moment until we need these. Same comment applies to all.

Oberon00 replied:
No we don't need them but it's a nice convenience feature. And since OpenTracing had similar features, it may ease porting of instrumentations from there to OpenTeletry. I have no strong opinion, but I think it's fine to keep them in this PR.

bogdandrutu replied:
Trying to limit the API for the moment. Any unnecessary method that we add now will cost us a lot to maintain/keep.

Oberon00 replied:
Ah I see, it's in the core API package. Which is IMHO good, since I think the semantic conventions should be seen as a core API. Maybe we should mark these methods `final` to limit the possibility of headaches in the future?

bogdandrutu replied:
I think the whole class should be final :)

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,83 @@
+   * @param attributeKey the attribute name/key
+   */
+  public BooleanAttributeSetter(String attributeKey) {
+    super();
```
why calling this?

bogdandrutu replied:
Ping on this one.

Oberon00 replied:
@kbrockhoff would be nice to fix this (though personally I wouldn't hold up the PR further just because of this superfluous line).

kbrockhoff replied:
This is fixed. I am not sure why it is showing as outdated.

bogdandrutu replied:
Am I looking to a wrong class https://github.com/open-telemetry/opentelemetry-java/pull/758/files#diff-55ba0be6445e26747fa46c33ffaffdaeR39?

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,83 @@
+
+/** Defines the behavior for a span attribute with boolean values. */
+@Immutable
+public class BooleanAttributeSetter {
```
Consider to use autovalue.

Oberon00 replied:
Why?

bogdandrutu replied:
To avoid checking for null the key and have autogenerated hash/equal if we need them :)

bogdandrutu replied:
Not a strong opinion, as I said is a consideration.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,83 @@
+   *
+   * @param attributeKey the attribute name/key
+   */
+  public BooleanAttributeSetter(String attributeKey) {
```
Consider to use a static factory method `create`. It is more flexible than the ctor.

kbrockhoff replied:
Changed to factory method

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,83 @@
+   */
+  public void set(Span span, boolean value) {
+    span.setAttribute(key(), value);
+  }
```
Another maybe useful helper would be `setToCurrent(boolean)` allows users to avoid getting the current `Span` :)

Oberon00 replied:
Hmm, that would kinda explode the interface if we also wanted things like `setParsedOrRawToCurrent`. In Python, there is a special sentinel value for CURRENT_SPAN that is of type Span. That could work for Java too.

We could solve this via a global helper method `curentSpan` that you would `import static`, then you can write `myAttr.set(currentSpan(), myBool)` which is not much longer than `myAttr.setToCurrent(myBool)`. Maybe such a method even exists already.

bogdandrutu replied:
We do have getCurrentSpan() :)

bogdandrutu replied:
So I am fine to not have the setToCurrent(). But I do want the `setParsedOrRaw` to be removed for the moment.

Oberon00 replied:
What about `trySetParsed`?

bogdandrutu replied:
Both, I think they can be built easily on top of just `Set` for the moment. Adding new methods is easier than removing in terms of backwards compatibility.

Oberon00 replied:
It's not possible to implement setParsedOrRaw with set, since you can only set integer values but not string values with that.

bogdandrutu replied:
Now that I see the behavior is even more confusing to set a string value using a int attribute setter.

kbrockhoff replied:
Removed requested methods

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/attributes/BooleanAttributeSetter.java:
```diff
@@ -0,0 +1,64 @@
+
+/** Defines the behavior for a span attribute with boolean values. */
+@Immutable
+public class BooleanAttributeSetter {
```
Make classes final to avoid unexpected inheritance behavior.

kbrockhoff replied:
Fixed
